cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/FlowControlService.java

```diff
@@ -0,0 +1,180 @@
+      TransactionRunner transactionRunner) {
+    this.metricsCollectionService = metricsCollectionService;
+    this.transactionRunner = transactionRunner;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("FlowControlService started.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("FlowControlService successfully shut down.");
+  }
+
+  /**
+   * Add a new in-flight launch request and return total number of launching and running programs.
+   *
+   * @param programRunId run id associated with the launch request
+   * @return total number of launching and running program runs.
+   */
+  public Counter addRequestAndGetCount(ProgramRunId programRunId, ProgramOptions programOptions,
+      ProgramDescriptor programDescriptor) throws Exception {
+    if (RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS) == -1) {
+      throw new Exception("None time-based UUIDs are not supported");
+    }
+
+    Counter counter = TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore store = AppMetadataStore.create(context);
+      store.recordProgramPending(programRunId,
+          programOptions.getArguments().asMap(),
+          programOptions.getUserArguments().asMap(),
+          programDescriptor.getArtifactId().toApiArtifactId());
+      int launchingCount = store.getFlowControlLaunchingCount();
+      int runningCount = store.getFlowControlRunningCount();
+      return new Counter(launchingCount, runningCount);
+    });
+    LOG.info("Added request with runId {}.", programRunId);
+    emitMetrics(Constants.Metrics.FlowControl.LAUNCHING_COUNT, counter.getLaunchingCount());
+
+    LOG.info(
+        "Counter has {} concurrent launching and {} running programs.",
+        counter.getLaunchingCount(),
+        counter.getRunningCount());
+    return counter;
+  }
+
+  /**
+   * Add a new in-flight launch request.
+   *
+   * @param programRunId run id associated with the launch request
+   *
+   * @return Returns the count of launching programs.
+   */
+  public int addRequest(ProgramRunId programRunId, ProgramOptions programOptions,
+      ProgramDescriptor programDescriptor) {
+    int launchingCount = TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore store = AppMetadataStore.create(context);
+      store.recordProgramPending(programRunId,
+          programOptions.getArguments().asMap(),
+          programOptions.getUserArguments().asMap(),
+          programDescriptor.getArtifactId().toApiArtifactId());
+      return store.getFlowControlLaunchingCount();
+    });
+    emitMetrics(Constants.Metrics.FlowControl.LAUNCHING_COUNT, launchingCount);
+    LOG.info("Added request with runId {}.", programRunId);
+    return launchingCount;
+  }
+
+  /**
+   * Get total number of launching and running programs.
+   *
+   * @return Counter with total number of launching and running program runs.
+   */
+  public Counter getCount() {
```

Comment:
nit: consider using plural (`getCounts`) or object name ('`getCounter`)

Reply:
Renamed to `getCounter()`. 

Similarly renamed `addRequestAndGetCount()` to `addRequestAndGetCounter()`